tortoisebeacon/tortoisebeacon.go

```diff
@@ -0,0 +1,652 @@
+
+	case DelayedMessage:
+		tb.Log.Debug("Received delayed ProposalMessage, epoch: %v, proposals: %v", m.Epoch(), m.Proposals())
+		return nil
+
+	case LateMessage:
+		tb.Log.Debug("Received late ProposalMessage, epoch: %v, proposals: %v", m.Epoch(), m.Proposals())
+		return nil
+
+	default:
+		return ErrUnknownMessageType
+	}
+}
+
+func (tb *TortoiseBeacon) handleVotingMessage(m VotingMessage) error {
+	epoch := m.Epoch()
+
+	tb.currentRoundsMu.Lock()
+	currentRound := tb.currentRounds[epoch]
+	tb.currentRoundsMu.Unlock()
+
+	pair := EpochRoundPair{
+		EpochID: epoch,
+		Round:   currentRound,
+	}
+
+	mt := tb.classifyMessage(m, epoch)
+	switch mt {
+	case TimelyMessage:
+		for _, hash := range m.VotesFor() {
+			tb.votesMu.Lock()
+			tb.votesFor[pair] = append(tb.votesFor[pair], hash)
+			tb.votesAgainst[pair] = append(tb.votesAgainst[pair], hash)
+			tb.votesMu.Unlock()
+		}
+
+		return nil
+
+	case DelayedMessage:
+		tb.Log.Debug("Received delayed VotingMessage, epoch: %v, proposals: %v, hash list: %v",
+			m.Epoch(), m.Round(), m.VotesFor())
+		return nil
+
+	case LateMessage:
+		tb.Log.Debug("Received late VotingMessage, epoch: %v, round: %v, hash list: %v",
+			m.Epoch(), m.Round(), m.VotesFor())
+		return nil
+
+	default:
+		return ErrUnknownMessageType
+	}
+}
+
+func (tb *TortoiseBeacon) classifyMessage(m Message, epoch types.EpochID) MessageType {
+	tb.currentRoundsMu.Lock()
+	currentRound := tb.currentRounds[epoch]
+	tb.currentRoundsMu.Unlock()
+
+	round := 0
+	if m, ok := m.(VotingMessage); ok {
+		round = m.Round()
+	}
+
+	switch {
+	case round <= currentRound-1:
+		return TimelyMessage
+	case round == currentRound-2:
+		return DelayedMessage
+	default:
+		return LateMessage
+	}
+}
+
+// For K rounds: In each round that lasts δ, wait for proposals to come in.
+func (tb *TortoiseBeacon) roundTicker(epoch types.EpochID) {
+	ticker := time.NewTicker(tb.networkDelta)
+	defer ticker.Stop()
+
+	tb.currentRoundsMu.Lock()
+	tb.currentRounds[epoch] = 0
+	tb.currentRoundsMu.Unlock()
+
+	// Round 0 is already happened at this point, starting from round 1.
+	// For next rounds,
+	// wait for δ time, and construct a message that points to all messages from previous round received by δ.
+	for i := 1; i < tb.lastPossibleRound(); i++ {
+		select {
+		case <-ticker.C:
+			tb.currentRoundsMu.Lock()
+			tb.currentRounds[epoch]++
+			round := tb.currentRounds[epoch]
+			tb.currentRoundsMu.Unlock()
+
+			if i == 1 {
+				// round 1, send hashed proposal
+				// create a voting message that references all seen proposals within δ time frame and send it
+				tb.timelyProposalsMu.RLock()
+				timelyProposals := tb.timelyProposalsList[epoch]
+				tb.timelyProposalsMu.RUnlock()
+
+				hashesFor := make([]types.Hash32, 0, len(timelyProposals))
+				for _, p := range timelyProposals {
+					hashesFor = append(hashesFor, hashATXList(p))
+				}
+
+				hashesAgainst := make([]types.Hash32, 0, len(timelyProposals))
+				for _, p := range timelyProposals {
```

Comment:
you are iterating the timelyProposals again, As I see it, since you don't store late messages there is not need for `VotesAgains` at all, only vote `For` proposals so that we can reach consensus about what are the correct values